### PR TITLE
Adding reactor implementation

### PR DESCRIPTION
**Pull request by marcingrzejszczak**

### CODE REVIEW

File: spring-cloud-sleuth-core/src/test/java/org/springframework/cloud/sleuth/instrument/reactor/SpanSubscriberTests.java [hunk 1]
```diff
@@ -58,7 +59,71 @@ public void setup() {
...
+				})
+				.map( d -> d + 1)
+				.subscribe(System.out::println);
+
+		then(this.tracer.getCurrentSpan()).isNull();
+		then(spanInOperation.get().getParents().get(0)).isEqualTo(span.getSpanId());
+		then(ExceptionUtils.getLastException()).isNull();
+	}
+
+	@Test public void should_not_trace_scalar_flows() {
```
Comment (adriancole) [inline spring-cloud-sleuth-core/src/test/java/org/springframework/cloud/sleuth/instrument/reactor/SpanSubscriberTests.java]: javadoc why?

File: spring-cloud-sleuth-core/src/test/java/org/springframework/cloud/sleuth/instrument/reactor/SpanSubscriberTests.java [hunk 2]
```diff
@@ -58,7 +59,71 @@ public void setup() {
 				.subscribe(System.out::println);
 
 		then(this.tracer.getCurrentSpan()).isNull();
-		then(spanInOperation.get().getTraceId()).isEqualTo(span.getTraceId());
+		then(spanInOperation.get().getParents().get(0)).isEqualTo(span.getSpanId());
+		then(ExceptionUtils.getLastException()).isNull();
+	}
+
+	@Test public void should_support_reactor_fusion_optimization() {
```
Comment (adriancole) [inline spring-cloud-sleuth-core/src/test/java/org/springframework/cloud/sleuth/instrument/reactor/SpanSubscriberTests.java]: javadoc what reactor fusion optimization means and why we are being careful to test it?

File: spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/reactor/TraceReactorAutoConfiguration.java [hunk 3]
```diff
@@ -31,21 +32,27 @@
...
 		@Autowired Tracer tracer;
 		@Autowired TraceKeys traceKeys;
 		@Autowired SpanNamer spanNamer;
 
 		@PostConstruct
 		public void setupHooks() {
-			Hooks.onNewSubscriber((pub, sub) ->
-					new SpanSubscriber(sub, sub.currentContext(), this.tracer, pub.toString()));
+			Hooks.onNewSubscriber((pub, sub) -> {
+				//do not trace fused flows or simple just/error/empty
```
Comment (adriancole) [inline spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/reactor/TraceReactorAutoConfiguration.java]: add a comment line of why

**Comment by marcingrzejszczak** [comment]
Yeah - we'll be adding the javadocs (ping @smaldini ). Currently I have some issues in making the build pass - that's why I filed this PR. But thanks for the review @adriancole

**Comment by marcingrzejszczak** [comment]
Done via https://github.com/spring-cloud/spring-cloud-sleuth/commit/7b11141cd7cce706c831018714d90502d192bb4c . Still we need to add the javadocs though.

**Pull request closed without merging**